numpy/ma/core.py

```diff
@@ -3248,10 +3234,22 @@ def __setitem__(self, indx, value):
                 _mask[indx] = mval
         elif not self._hardmask:
             # Unshare the mask if necessary to avoid propagation
+            # We want to remove the unshare logic from this place in the
+            # future. Note that _sharedmask has lots of false positives.
             if not self._isfield:
-                _oldsharedmask = self._oldsharedmask
+                if self._sharedmask and not (
+                        # No one else _mask, self._mask + 1 holds a reference
+                        # and the array owns its own data
+                        (sys.getrefcount(_mask) == 3) and _mask.flags.owndata):
```

Comment:
1 for getrefcount, one for the attribute and one for the _mask local variable. Anything more (or not owndata) might mean someone else is using the same mask. Made the comment more understandable (will push when I got a 0.1 version of the release notes text).